File: backend/app/schemas/labs.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel


class ApiSchema(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class LabListItem(ApiSchema):
    id: str
    name: str
    professor_name: str
    university: str
    department: str
    field: str
    summary: str | None
    keywords: list[str]
    homepage_url: str | None
    updated_at: datetime
    recommendation_score: int | None
    is_favorite: bool


class LabSearchResponse(ApiSchema):
    items: list[LabListItem]
    page: int = Field(ge=1)
    page_size: int = Field(ge=1)
    total: int = Field(ge=0)


class SimilarLabsResponse(ApiSchema):
    items: list[LabListItem]


class LabFactResponse(ApiSchema):
    model_config = ConfigDict(alias_generator=to_camel, from_attributes=True, populate_by_name=True)

    fact_type: str
    value_text: str | None
    value_number: int | None
    audience: str | None
    origin: str
    source_url: str | None
    source_checked_at: datetime | None


class PaperResponse(ApiSchema):
    id: str
    title: str
    venue: str
    published_year: int
    abstract: str | None
    summary: str | None
    keywords: list[str]
    paper_url: str | None
    source_url: str | None
    source_checked_at: datetime | None


class LabDetail(LabListItem):
    location: str | None
    contact_email: str | None
    source_url: str | None
    source_checked_at: datetime | None
    facts: list[LabFactResponse]
    papers: list[PaperResponse]

File: frontend/src/lib/api/labs.ts
import { queryOptions } from "@tanstack/react-query";
import { apiFetch } from "./client";

export class ApiError extends Error {
  constructor(
    message: string,
    readonly status: number,
  ) {
    super(message);
    this.name = "ApiError";
  }
}

export type LabListItem = {
  id: string;
  name: string;
  professorName: string;
  university: string;
  department: string;
  field: string;
  summary: string | null;
  keywords: string[];
  homepageUrl: string | null;
  updatedAt: string;
  recommendationScore: number | null;
  isFavorite: boolean;
};

export type LabFact = {
  factType: string;
  valueText: string | null;
  valueNumber: number | null;
  audience: string | null;
  origin: string;
  sourceUrl: string | null;
  sourceCheckedAt: string | null;
};

export type LabPaper = {
  id: string;
  title: string;
  venue: string;
  publishedYear: number;
  abstract: string | null;
  summary: string | null;
  keywords: string[];
  paperUrl: string | null;
  sourceUrl: string | null;
  sourceCheckedAt: string | null;
};

export type LabDetail = LabListItem & {
  location: string | null;
  contactEmail: string | null;
  sourceUrl: string | null;
  sourceCheckedAt: string | null;
  facts: LabFact[];
  papers: LabPaper[];
};

export type LabSearchParams = {
  q?: string;
  department?: string;
  fields?: string[];
  sort?: "score" | "recent";
  page?: number;
  pageSize?: number;
};

export type LabSearchResponse = {
  items: LabListItem[];
  page: number;
  pageSize: number;
  total: number;
};

async function request<T>(path: string): Promise<T> {
  const response = await apiFetch(path);
  if (!response.ok) {
    throw new ApiError("Unable to load lab data.", response.status);
  }
  return response.json() as Promise<T>;
}

export async function getLabs(params: LabSearchParams = {}): Promise<LabSearchResponse> {
  const query = new URLSearchParams();
  if (params.q) query.set("q", params.q);
  if (params.department) query.set("department", params.department);
  params.fields?.forEach((field) => query.append("field", field));
  query.set("sort", params.sort ?? "recent");
  query.set("page", String(params.page ?? 1));
  query.set("page_size", String(params.pageSize ?? 20));
  return request<LabSearchResponse>(`/labs?${query.toString()}`);
}

export async function getLab(labId: string): Promise<LabDetail> {
  return request<LabDetail>(`/labs/${encodeURIComponent(labId)}`);
}

export async function getSimilarLabs(labId: string): Promise<LabListItem[]> {
  const response = await request<{ items: LabListItem[] }>(
    `/labs/${encodeURIComponent(labId)}/similar?limit=3`,
  );
  return response.items;
}

export const labsQueryOptions = (params: LabSearchParams) =>
  queryOptions({ queryKey: ["labs", params], queryFn: () => getLabs(params) });

export const labQueryOptions = (labId: string) =>
  queryOptions({ queryKey: ["labs", labId], queryFn: () => getLab(labId) });

export const similarLabsQueryOptions = (labId: string) =>
  queryOptions({ queryKey: ["labs", labId, "similar"], queryFn: () => getSimilarLabs(labId) });
